src/indexer.c

```diff
@@ -33,6 +34,39 @@ void IndexInit() {
     tsLabelIndex = RedisModule_CreateDict(NULL);
 }
 
+static void *NullDefrag(__unused RedisModuleDefragCtx *ctx,
+                        void *data,
+                        __unused unsigned char *key,
+                        __unused size_t keylen) {
+    return data;
+}
+
+static void *DefragIndexLeaf(RedisModuleDefragCtx *ctx,
+                             void *data,
+                             __unused unsigned char *key,
+                             __unused size_t keylen) {
+    static RedisModuleString *seekTo = NULL;
+    return (void *)defragDict(ctx, (RedisModuleDict *)data, NullDefrag, &seekTo);
+}
+
+int DefragIndex(RedisModuleDefragCtx *ctx) {
+    static RedisModuleString *seekTo = NULL;
+    static RedisModuleDict **index = &labelsIndex;
+
+    // can only defrag one index at a time
+    *index = defragDict(ctx, *index, DefragIndexLeaf, &seekTo);
+    if (seekTo != NULL) { // defrag paused
+        return DefragStatus_Paused;
+    }
+
+    index = (index == &labelsIndex) ? &tsLabelIndex : &labelsIndex;
+    if (index == &labelsIndex) { // defragged both indexes, done
+        return DefragStatus_Finished;
+    }
+
+    return DefragIndex(ctx);
```

Comment:
Lets make it easier to to understand and avoid recursion.

Reply:
The recursion may be called either zero or one times. I think adding a comment explaining the behavior is much easier to explain than duplicating the code with an optional fallthrough between the two sections.

There should be no extra stack frame used due to tail recursion. Both gcc and clang both have very strong tail recursion heuristics. If you prefer, I can add `__attribute__((__flatten__))` to the signature to guarantee the lack of recursion in the generated asm. But that will expand the generated asm instead of jumping back to the beginning of the same stack frame.